File: c++_solutions/205.cpp
class Solution {
public:
    bool isIsomorphic(string s, string t) {
        unordered_map<char,char> s2t , t2s ; 
        auto sitr = s.begin() ;
        auto titr = t.begin() ; 
        while (sitr != s.end() && titr != t.end()){
            auto st_char = s2t.find(*sitr);
            auto ts_char = t2s.find(*titr);
            if (st_char != s2t.end() && ts_char != t2s.end()){
                if ((st_char->second != *titr) || (ts_char->second != *sitr)){
                    return false;
                }
            }
            else if (st_char == s2t.end() && ts_char == t2s.end()){
                s2t[*sitr] = *titr ;
                t2s[*titr] = *sitr ;
            }
            else{
                return false;
            }
            sitr++;
            titr++;
        }
        return true ;
    }
};

File: rust_solutions/205.rs
impl Solution {
    pub fn is_isomorphic(s: String, t: String) -> bool {
        use std::collections::HashMap ; 
        let mut s2t : HashMap<char,char> = HashMap::new();
        let mut t2s : HashMap<char,char> = HashMap::new();
        let mut sitr = s.chars() ;
        let mut titr = t.chars() ;
        while let (Some(s_char),Some(t_char)) = (sitr.next(),titr.next()){
            match (s2t.get(&s_char),t2s.get(&t_char)){
                (Some(&st_char),Some(&ts_char)) => {
                    if st_char != t_char || ts_char != s_char {
                        return false
                    }
                },
                (None,None) => {
                    s2t.insert(s_char,t_char);
                    t2s.insert(t_char,s_char);
                },
                _ => return false
            }
        }
        true
    }
}
